File: tilskuerstats/verify.py
"""Tjekker at et udkast kun bruger tal, der findes i værktøjsresultaterne.

Det er agentens vigtigste værn mod opfundne tal: modellen må gerne vælge,
hvilke tal der er interessante, og runde dem af, men den må ikke selv
finde på et tal. Fejler tjekket, sendes fejlen tilbage som tool_result, og
modellen får chancen for at rette udkastet.
"""
from __future__ import annotations

import re
from typing import Any, Iterable

# "12.456", "12,5", "1.234,5", "2024", "3"
NUMBER_RE = re.compile(r"\d+(?:[.,]\d+)*")


def parse_danish_number(token: str) -> tuple[float, int]:
    """Returnerer (værdi, antal decimaler). Punktum er tusindtalsseparator,
    komma er decimaltegn, som på dansk."""
    if "," in token:
        whole, _, frac = token.rpartition(",")
        whole = whole.replace(".", "")
        return float(f"{whole}.{frac}"), len(frac)
    parts = token.split(".")
    if len(parts) > 1 and all(len(p) == 3 for p in parts[1:]):
        return float("".join(parts)), 0
    # "23.4" i engelsk notation (kan optræde i værktøjsoutput)
    return float(token), len(parts[1]) if len(parts) == 2 else 0


def numbers_in_text(text: str) -> list[tuple[str, float, int]]:
    return [(m.group(0), *parse_danish_number(m.group(0))) for m in NUMBER_RE.finditer(text)]


def numbers_in_result(value: Any) -> set[float]:
    """Alle tal i et (JSON-)værktøjsresultat, også dem inde i strenge som
    datoer ('2025-09-28T16:00') og resultater ('2-1')."""
    found: set[float] = set()
    if isinstance(value, bool) or value is None:
        return found
    if isinstance(value, (int, float)):
        found.add(abs(float(value)))
    elif isinstance(value, str):
        # Engelsk notation i værktøjsstrenge: split på ikke-tal-tegn
        found.update(float(t) for t in re.findall(r"\d+(?:\.\d+)?", value))
    elif isinstance(value, dict):
        for k, v in value.items():
            found |= numbers_in_result(k) | numbers_in_result(v)
    elif isinstance(value, (list, tuple)):
        for v in value:
            found |= numbers_in_result(v)
    return found


def is_grounded(value: float, decimals: int, sources: Iterable[float]) -> bool:
    """Et tal er grundet, hvis en kilde afrundet til samme præcision giver tallet."""
    return any(round(s, decimals) == value for s in sources)


def ungrounded_numbers(text: str, results: Iterable[Any]) -> list[str]:
    sources = set()
    for r in results:
        sources |= numbers_in_result(r)
    return [tok for tok, value, dec in numbers_in_text(text) if not is_grounded(value, dec, sources)]

File: src/insights/verify.ts
// Tjekker at et udkast kun bruger tal, der står i agentens værktøjsresultater.
//
// Det er agentens vigtigste værn mod opfundne tal: modellen må gerne vælge,
// hvilke tal der er interessante, og runde dem af, men den må ikke selv
// finde på et. Fejler tjekket, sendes fejlen tilbage til modellen som et
// tool_result med is_error, så den kan rette udkastet.

// "12.456", "12,5", "1.234,5", "2024", "3"
const NUMBER_RE = /\d+(?:[.,]\d+)*/g;

// Id'er (cuid) og slugs indeholder tilfældige cifre, som ellers ville gøre
// vilkårlige tal "godkendte".
const IGNORED_KEYS = /(^id$|Id$|_id$|slug$)/;

/** Dansk talformat: punktum er tusindtalsseparator, komma er decimaltegn. */
export function parseDanishNumber(token: string): { value: number; decimals: number } {
  if (token.includes(",")) {
    const i = token.lastIndexOf(",");
    const whole = token.slice(0, i).replaceAll(".", "");
    const frac = token.slice(i + 1);
    return { value: Number(`${whole}.${frac}`), decimals: frac.length };
  }
  const parts = token.split(".");
  if (parts.length > 1 && parts.slice(1).every((p) => p.length === 3)) {
    return { value: Number(parts.join("")), decimals: 0 };
  }
  // Engelsk notation ("23.4") kan også optræde
  return { value: Number(token), decimals: parts.length === 2 ? parts[1].length : 0 };
}

/** Alle tal i et værktøjsresultat, også dem inde i strenge som datoer og resultater. */
export function numbersInResult(value: unknown, into = new Set<number>()): Set<number> {
  if (typeof value === "number") {
    into.add(Math.abs(value));
  } else if (typeof value === "string") {
    // Hvert ciffer-løb for sig ("20.07.2025" -> 20, 7, 2025) plus hele danske
    // tal ("1.200 pladser" -> 1200), så begge skrivemåder kan godkendes.
    for (const t of value.match(/\d+/g) ?? []) into.add(Number(t));
    for (const t of value.match(NUMBER_RE) ?? []) {
      const n = parseDanishNumber(t).value;
      if (Number.isFinite(n)) into.add(n);
    }
  } else if (Array.isArray(value)) {
    for (const v of value) numbersInResult(v, into);
  } else if (value && typeof value === "object") {
    for (const [k, v] of Object.entries(value)) {
      if (!IGNORED_KEYS.test(k)) numbersInResult(v, into);
    }
  }
  return into;
}

function roundTo(value: number, decimals: number): number {
  const f = 10 ** decimals;
  return Math.round(value * f) / f;
}

/** Returnerer de tal i `text`, der ikke kan findes (evt. afrundet) i `results`. */
export function ungroundedNumbers(text: string, results: unknown[]): string[] {
  const sources = new Set<number>();
  for (const r of results) numbersInResult(r, sources);
  const bad: string[] = [];
  for (const token of text.match(NUMBER_RE) ?? []) {
    const { value, decimals } = parseDanishNumber(token);
    const grounded = [...sources].some((s) => roundTo(s, decimals) === value);
    if (!grounded) bad.push(token);
  }
  return bad;
}
